README.md
# cargo-fund

Discover funding links for your project's dependencies.

[![Crates.io][crates-badge]][crates-url]
[![CircleCI][circleci-badge]][circleci-url]

[crates-badge]: https://img.shields.io/crates/v/cargo-fund.svg
[crates-url]: https://crates.io/crates/cargo-fund
[circleci-badge]: https://img.shields.io/circleci/build/github/acfoltzer/cargo-fund/develop
[circleci-url]: https://circleci.com/gh/acfoltzer/cargo-fund

## Installation

To install `cargo-fund`, use `cargo`:

```sh
$ cargo install cargo-fund
```

### Github API token

`cargo-fund` retrieves funding links for any dependencies with a Github URL in its
`[package.repository]` field. To retrieve this information, you must provide a valid Github API
token in the `CARGO_FUND_GITHUB_API_TOKEN` environment variable or the `--github-api-token`
command-line argument. To generate this token, go to <https://github.com/settings/tokens> and create
a token with the `public_repo` and `user` scopes.

## Usage

Run `cargo fund` in your workspace to print funding links. For example:

```text
$ CARGO_FUND_GITHUB_API_TOKEN=... cargo fund
/path/to/cargo-fund (found funding links for 16 out of 138 dependencies)
├─┬─ https://www.buymeacoffee.com/dannyguo
│ ├─ https://www.paypal.me/DannyGuo
│ └─ https://ko-fi.com/dannyguo
│    └─ strsim 0.8.0
├─── https://github.com/sponsors/XAMPPRocky
│    └─ remove_dir_all 0.5.2
├─── https://github.com/sponsors/dtolnay
│    ├─ anyhow 1.0.28
│    ├─ dtoa 0.4.5
│    ├─ itoa 0.4.5
│    ├─ proc-macro-hack 0.5.15
│    ├─ proc-macro-nested 0.1.4
│    ├─ quote 1.0.3
│    ├─ ryu 1.0.4
│    └─ syn 1.0.18
└─── https://github.com/sponsors/seanmonstar
     ├─ httparse 1.3.4
     ├─ num_cpus 1.13.0
     ├─ reqwest 0.10.4
     ├─ try-lock 0.2.2
     ├─ unicase 2.6.0
     └─ want 0.3.0
```

## Including your sponsorship info

`cargo-fund` uses the Github API to get the available funding links for crates. To ensure your
crate's information appears:

1. Make sure that the `[package.repository]` in your `Cargo.toml` contains a valid Github URL.
2. Add your funding information to [`.github/FUNDING.yml`][funding-yml] in your repository.

Currently, Github is the only source of funding information, but please open an issue if you know of
any other structured sources of funding information.

[funding-yml]: https://help.github.com/en/github/administering-a-repository/displaying-a-sponsor-button-in-your-repository

#### License

<sup>
Licensed under either of <a href="LICENSE-APACHE">Apache License, Version
2.0</a> or <a href="LICENSE-MIT">MIT license</a> at your option.
</sup>

<br>

<sub>
Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in this crate by you, as defined in the Apache-2.0 license, shall
be dual licensed as above, without any additional terms or conditions.
</sub>


CHANGELOG.md
# 0.2.3

## Changed

- Updated to `clap` 4.

## Fixed

- Updated some dependencies to clear Dependabot warnings. None of the warnings were applicable to how `cargo-fund` used those dependencies directly.

# 0.2.2

## Fixed

- Updated some dependencies to clear Dependabot warnings. None of the warnings were applicable to how `cargo-fund` used those dependencies.

# 0.2.1

## Fixed

- ([#8](https://github.com/acfoltzer/cargo-fund/issues/8)) `cargo-fund` will now try to prepend an `https://` scheme to bare relative URLs, and failure to parse a URL will lead to that entry being ignored rather than crashing.

# 0.2.0

## Changed

- ([#5](https://github.com/acfoltzer/cargo-fund/pull/5)) The program now looks for the Github API token in the `CARGO_FUND_GITHUB_API_TOKEN` environment variable rather than `GITHUB_API_TOKEN` in order to better support privilege separation.
- ([#6](https://github.com/acfoltzer/cargo-fund/pull/6)) Increased timeout on HTTP client operations to 60 seconds.

# 0.1.0

Initial release.


tests/test.rs
use std::path::Path;
use std::process::Command;

fn sanitize_stderr(bytes: &[u8]) -> String {
    let stderr = std::str::from_utf8(bytes).expect("stderr is valid UTF-8");
    // since we're running a bunch of cargo commands back to back, this message can sometimes show
    // up in the test output
    stderr.replace("    Blocking waiting for file lock on package cache\n", "")
}

#[test]
fn client_package_output_expected() {
    let root = Path::new(env!("CARGO_MANIFEST_DIR"));
    let client_package = root.join("tests").join("client-package");
    let expected = format!(
        "{} (found funding links for 1 out of 3 dependencies)
──┬─ https://acfoltzer.net/bare_relative_link
  ├─ https://www.acfoltzer.net/
  ├─ https://www.acfoltzer.net/another_url
  ├─ https://issuehunt.io/r/acfoltzer
  ├─ https://ko-fi.com/acfoltzer
  ├─ https://liberapay.com/acfoltzer
  └─ https://patreon.com/acfoltzer
     └─ funding-test 0.1.0\n",
        client_package.display()
    );
    let exe = Path::new(env!("CARGO_BIN_EXE_cargo-fund"));
    let output = Command::new(exe)
        .current_dir(client_package)
        .arg("fund")
        .env(
            "CARGO_FUND_GITHUB_API_TOKEN",
            std::env::var_os("VALID_CARGO_FUND_GITHUB_API_TOKEN").unwrap(),
        )
        .output()
        .expect("cargo-fund runs");

    assert!(output.status.success());
    let stdout = std::str::from_utf8(&output.stdout).expect("stdout is valid UTF-8");
    assert_eq!(stdout, expected, "stdout matches");
    assert_eq!(&sanitize_stderr(&output.stderr), "", "stderr matches");
}

#[test]
fn missing_token() {
    let root = Path::new(env!("CARGO_MANIFEST_DIR"));
    let expected =
        "Error: Github API token must be provided through the CARGO_FUND_GITHUB_API_TOKEN \
         environment variable or the --github-api-token flag.\n";
    let exe = Path::new(env!("CARGO_BIN_EXE_cargo-fund"));
    let output = Command::new(exe)
        .current_dir(root.join("tests").join("client-package"))
        .arg("fund")
        // not necessary for CI, but makes local testing easier
        .env_remove("CARGO_FUND_GITHUB_API_TOKEN")
        .output()
        .expect("cargo-fund runs");
    assert!(!output.status.success());
    assert_eq!(&output.stdout, b"", "stdout matches");
    assert_eq!(&sanitize_stderr(&output.stderr), expected, "stderr matches");
}

#[test]
fn invalid_token() {
    let root = Path::new(env!("CARGO_MANIFEST_DIR"));
    let expected =
        "Error: Invalid Github API token. Create a token with the `public_repo` and `user` scopes \
         at https://github.com/settings/tokens.\n";
    let exe = Path::new(env!("CARGO_BIN_EXE_cargo-fund"));
    let mut token = std::env::var("VALID_CARGO_FUND_GITHUB_API_TOKEN").unwrap();
    // remove a character to invalidate the token
    token.pop();

    let output = Command::new(exe)
        .current_dir(root.join("tests").join("client-package"))
        .arg("fund")
        .env("CARGO_FUND_GITHUB_API_TOKEN", token)
        .output()
        .expect("cargo-fund runs");
    assert!(!output.status.success());
    assert_eq!(&output.stdout, b"", "stdout matches");
    assert_eq!(&sanitize_stderr(&output.stderr), expected, "stderr matches");
}

#[test]
fn insufficient_scopes() {
    let root = Path::new(env!("CARGO_MANIFEST_DIR"));
    let expected = "Error: Insufficient Github API token scopes. Modify your token to include the \
                    `public_repo` and `user` scopes at https://github.com/settings/tokens.\n";
    let exe = Path::new(env!("CARGO_BIN_EXE_cargo-fund"));
    let output = Command::new(exe)
        .current_dir(root.join("tests").join("client-package"))
        .arg("fund")
        .env(
            "CARGO_FUND_GITHUB_API_TOKEN",
            std::env::var_os("INSUFFICIENT_SCOPES_CARGO_FUND_GITHUB_API_TOKEN").unwrap(),
        )
        .output()
        .expect("cargo-fund runs");
    assert!(!output.status.success());
    assert_eq!(&output.stdout, b"", "stdout matches");
    assert_eq!(&sanitize_stderr(&output.stderr), expected, "stderr matches");
}


src/args.rs
//! Adapted from the `cargo_tree::args` module.

use clap::{ArgAction, Parser};
use serde::Deserialize;
use std::path::PathBuf;

#[derive(Deserialize, Debug)]
pub struct Env {
    #[serde(rename = "cargo_fund_github_api_token")]
    pub github_api_token: Option<String>,
}

#[derive(Parser)]
#[clap(bin_name = "cargo")]
pub enum Opts {
    #[clap(name = "fund", dont_collapse_args_in_usage = true)]
    /// Display funding links for workspace dependencies
    Fund(Args),
}

#[derive(Parser)]
pub struct Args {
    /// Github API token, which must have the scope `public_repo`. This option overrides the token
    /// provided in the `CARGO_FUND_GITHUB_API_TOKEN` environment variable.
    #[clap(long = "github-api-token", value_name = "TOKEN")]
    pub github_api_token: Option<String>,
    #[clap(long = "manifest-path", value_name = "PATH", value_parser)]
    /// Path to Cargo.toml
    pub manifest_path: Option<PathBuf>,
    #[clap(long = "verbose", short = 'v', action = ArgAction::Count)]
    /// Use verbose output (-vv very verbose/build.rs output)
    pub verbose: u8,
    #[clap(long = "quiet", short = 'q')]
    /// No output printed to stdout other than the funding information
    pub quiet: bool,
    #[clap(long = "color", value_name = "WHEN")]
    /// Coloring: auto, always, never
    pub color: Option<String>,
    #[clap(short = 'Z', value_name = "FLAG")]
    /// Unstable (nightly-only) flags to Cargo
    pub unstable_flags: Vec<String>,
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn verify_cli() {
        use clap::CommandFactory;
        Opts::command().debug_assert();
    }
}


src/github.rs
use super::{globals, Link, LinkSource, Platform};
use anyhow::{anyhow, bail, Error};
use cargo_metadata::PackageId;
use http::{StatusCode, Uri};
use std::collections::{HashMap, HashSet};
use std::convert::TryFrom;
use std::fmt::Write;
use tracing::{debug, info, trace, warn};

const GITHUB_TOKEN_HELP: &str = "Invalid Github API token. \
Create a token with the `public_repo` and `user` scopes at https://github.com/settings/tokens.";

const GITHUB_TOKEN_SCOPES_HELP: &str = "Insufficient Github API token scopes. \
Modify your token to include the `public_repo` and `user` scopes at https://github.com/settings/tokens.";

#[derive(Clone, Debug, Hash, Eq, PartialEq, Ord, PartialOrd)]
pub(crate) enum GithubLinkSource {
    Repo { owner: String, name: String },
    Owner { owner: String },
}

impl GithubLinkSource {
    fn owner(&self) -> &str {
        match self {
            GithubLinkSource::Repo { owner, .. } => owner,
            GithubLinkSource::Owner { owner, .. } => owner,
        }
    }
}

pub(crate) fn try_get_sources(uri: Uri) -> Result<Vec<LinkSource>, Error> {
    let mut path_components = uri.path().split("/").skip(1).take(2);
    let owner = path_components.next();
    let name = path_components.next();
    if let (Some(owner), Some(name)) = (owner, name) {
        let name = name.trim_end_matches(".git");
        Ok(vec![
            LinkSource::Github(GithubLinkSource::Repo {
                owner: owner.to_string(),
                name: name.to_string(),
            }),
            LinkSource::Github(GithubLinkSource::Owner {
                owner: owner.to_string(),
            }),
        ])
    } else {
        bail!("not a full Github URI: {}", uri)
    }
}

pub(crate) async fn resolve_github_links(
    source_map: &HashMap<LinkSource, HashSet<PackageId>>,
    resolved: &mut HashMap<PackageId, HashSet<Link>>,
) -> Result<(), Error> {
    #[derive(Clone, Debug, Eq, PartialEq, Hash)]
    enum Alias {
        Repo(String),
        Owner(String),
    }
    let mut query_map = HashMap::new();
    let mut gensym = 0usize;
    let mut query = "query FundingLinks {".to_string();
    for (source, pkgs) in source_map {
        let alias = format!("_{}", gensym);
        gensym += 1;
        // allow this pattern even though we have no other `LinkSource` variants yet
        #[allow(irrefutable_let_patterns)]
        let source = if let LinkSource::Github(source) = source {
            source
        } else {
            continue;
        };
        match &source {
            GithubLinkSource::Repo { owner, name } => {
                writeln!(
                    &mut query,
                    "
{}: repository(owner: {:?}, name: {:?}) {{
  fundingLinks {{
    platform
    url
  }}
}}",
                    alias, owner, name,
                )
                .unwrap();
                query_map.insert(Alias::Repo(alias), (source, pkgs));
            }
            GithubLinkSource::Owner { owner } => {
                writeln!(
                    &mut query,
                    "
{}: repositoryOwner(login: {:?}) {{
  ... on Organization {{
    sponsorsListing {{
      id
    }}
  }}
  ... on User {{
    sponsorsListing {{
      id
    }}
  }}
}}
",
                    alias, owner
                )
                .unwrap();
                query_map.insert(Alias::Owner(alias), (source, pkgs));
            }
        }
    }
    writeln!(
        &mut query,
        "
  rateLimit {{
    cost
    remaining
  }}
}}"
    )
    .unwrap();

    let query = serde_json::json!({ "query": query });

    let req = globals()
        .client
        .post("https://api.github.com/graphql")
        .bearer_auth(&globals().github_api_token)
        .json(&query);

    trace!("sending Github GraphQL query");

    let resp = req.send().await?;

    trace!("received Github GraphQL query response");

    match resp.status() {
        StatusCode::OK => (),
        StatusCode::UNAUTHORIZED => bail!(GITHUB_TOKEN_HELP),
        status => bail!("Github API returned unexpected status: {}", status),
    }

    trace!("deserializing Github response JSON");

    let res: serde_json::Value = resp.json().await?;

    trace!("deserialized Github response JSON");

    if let serde_json::Value::Array(errors) = &res["errors"] {
        for error in errors {
            let message = error["message"]
                .as_str()
                .ok_or_else(|| anyhow!("Malformed Github API response"))?;
            if let serde_json::Value::String(ty) = &error["type"] {
                match ty.as_str() {
                    "INSUFFICIENT_SCOPES" => bail!(GITHUB_TOKEN_SCOPES_HELP),
                    "NOT_FOUND" => {
                        info!("{}", message);
                        continue;
                    }
                    _ => {
                        eprintln!("{}", error);
                        bail!("Github API response contained error: {}", message)
                    }
                }
            } else {
                bail!("Malformed Github API response");
            }
        }
    }

    for (alias, (source, pkgs)) in query_map {
        trace!("processing {:?}, {:?}", alias, source);
        match alias {
            Alias::Repo(alias) => {
                if let serde_json::Value::Array(links) = &res["data"][alias]["fundingLinks"] {
                    for link in links {
                        trace!("processing {:?}", link);
                        let platform = link["platform"]
                            .as_str()
                            .ok_or_else(|| anyhow!("Malformed Github API response"))?;
                        let uri = link["url"]
                            .as_str()
                            .ok_or_else(|| anyhow!("Malformed Github API response"))?;
                        let link = match Link::try_from((platform, uri)) {
                            Ok(link) => link,
                            Err(e) => {
                                warn!(
                                    platform = %platform,
                                    uri = %uri,
                                    "could not parse Github funding links; skipping: {}",
                                    e
                                );
                                continue;
                            }
                        };
                        for pkg in pkgs.iter() {
                            resolved
                                .entry(pkg.clone())
                                .or_insert_with(HashSet::new)
                                .insert(link.clone());
                        }
                    }
                } else {
                    // no result, probably indicates an invalid or private repo
                    continue;
                }
            }
            Alias::Owner(alias) => {
                if let serde_json::Value::Null = res["data"][alias]["sponsorsListing"] {
                    continue;
                } else {
                    let uri: http::Uri =
                        match format!("https://github.com/sponsors/{}", source.owner()).parse() {
                            Ok(link) => link,
                            Err(e) => {
                                warn!(
                                    owner = %source.owner(),
                                    "could not create valid owner sponsor link; skipping: {}",
                                    e
                                );
                                continue;
                            }
                        };
                    for pkg in pkgs {
                        resolved
                            .entry(pkg.clone())
                            .or_insert_with(HashSet::new)
                            .insert(Link {
                                platform: Platform::Github,
                                uri: uri.clone(),
                            });
                    }
                }
            }
        }
    }

    debug!("finished resolving Github links");

    Ok(())
}


src/metadata.rs
//! Adapted from the `cargo_tree::metadata` module.

use crate::args::Args;
use anyhow::{anyhow, Context, Error};
use cargo_metadata::Metadata;
use std::env;
use std::ffi::OsString;
use std::process::{Command, Stdio};

pub fn get(args: &Args) -> Result<Metadata, Error> {
    let cargo = env::var_os("CARGO").unwrap_or_else(|| OsString::from("cargo"));

    let mut command = Command::new(cargo);
    command.arg("metadata").arg("--format-version").arg("1");

    if args.quiet {
        command.arg("-q");
    }

    if let Some(path) = &args.manifest_path {
        command.arg("--manifest-path").arg(path);
    }

    for _ in 0..args.verbose {
        command.arg("-v");
    }

    if let Some(color) = &args.color {
        command.arg("--color").arg(color);
    }

    for flag in &args.unstable_flags {
        command.arg("-Z").arg(flag);
    }

    let output = output(&mut command, "cargo metadata")?;

    serde_json::from_str(&output).context("error parsing cargo metadata output")
}

fn output(command: &mut Command, job: &str) -> Result<String, Error> {
    let output = command
        .stderr(Stdio::inherit())
        .output()
        .with_context(|| format!("error running {}", job))?;

    if !output.status.success() {
        return Err(anyhow!("{} returned {}", job, output.status));
    }

    String::from_utf8(output.stdout).with_context(|| format!("error parsing {} output", job))
}


src/main.rs
//! `cargo fund`: find out how to financially support the authors of your dependencies.
//!
//! For example, when run on this project, `cargo fund` produces:
//!
//! ```text
//! % CARGO_FUND_GITHUB_API_TOKEN=... cargo fund
//! $HOME/cargo-fund (found funding links for 16 out of 138 dependencies)
//! ├─┬─ https://www.buymeacoffee.com/dannyguo
//! │ ├─ https://www.paypal.me/DannyGuo
//! │ └─ https://ko-fi.com/dannyguo
//! │    └─ strsim 0.8.0
//! ├─── https://github.com/sponsors/XAMPPRocky
//! │    └─ remove_dir_all 0.5.2
//! ├─── https://github.com/sponsors/dtolnay
//! │    ├─ anyhow 1.0.28
//! │    ├─ dtoa 0.4.5
//! │    ├─ itoa 0.4.5
//! │    ├─ proc-macro-hack 0.5.15
//! │    ├─ proc-macro-nested 0.1.4
//! │    ├─ quote 1.0.3
//! │    ├─ ryu 1.0.4
//! │    └─ syn 1.0.18
//! └─── https://github.com/sponsors/seanmonstar
//!      ├─ httparse 1.3.4
//!      ├─ num_cpus 1.13.0
//!      ├─ reqwest 0.10.4
//!      ├─ try-lock 0.2.2
//!      ├─ unicase 2.6.0
//!      └─ want 0.3.0
//! ```
//!
//! # Github API token
//!
//! `cargo fund` retrieves funding links for any dependencies with a Github URL in its
//! `[package.repository]` field. To retrieve this information, you must provide a valid Github API
//! token in the `CARGO_FUND_GITHUB_API_TOKEN` environment variable or the `--github-api-token` command-line
//! argument. To generate this token, go to <https://github.com/settings/tokens> and create a token
//! with the `public_repo` and `user` scopes.
use crate::args::Opts;
use anyhow::{anyhow, bail, Error};
use cargo_metadata::{Metadata, Package, PackageId};
use clap::Parser;
use lazy_static::lazy_static;
use parking_lot::{MappedRwLockReadGuard, RwLock, RwLockReadGuard};
use std::collections::{BTreeMap, BTreeSet, HashMap, HashSet};
use std::convert::{TryFrom, TryInto};

mod args;
mod github;
mod metadata;

lazy_static! {
    static ref GLOBALS: RwLock<Option<Globals>> = RwLock::new(None);
}

struct Globals {
    github_api_token: String,
    client: reqwest::Client,
}

fn initialize_globals(env: &args::Env, args: &args::Args) -> Result<(), Error> {
    tracing_subscriber::fmt::fmt()
        .with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
        .with_writer(std::io::stderr)
        .init();
    let github_api_token = if let Some(token) = args
        .github_api_token
        .as_ref()
        .or(env.github_api_token.as_ref())
    {
        token
    } else {
        bail!(
            "Github API token must be provided through the CARGO_FUND_GITHUB_API_TOKEN environment \
             variable or the --github-api-token flag."
        );
    };
    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(60))
        .user_agent(concat!(
            env!("CARGO_PKG_NAME"),
            "/",
            env!("CARGO_PKG_VERSION")
        ))
        .build()?;
    *GLOBALS.write() = Some(Globals {
        github_api_token: github_api_token.to_string(),
        client,
    });
    Ok(())
}

fn globals() -> MappedRwLockReadGuard<'static, Globals> {
    RwLockReadGuard::map(GLOBALS.read(), |o| {
        o.as_ref().expect("globals must be initialized first")
    })
}

#[derive(Clone, Debug, Hash, Eq, PartialEq, Ord, PartialOrd)]
enum LinkSource {
    Github(github::GithubLinkSource),
}

#[derive(Clone, Debug, Hash, Eq, PartialEq, Ord, PartialOrd)]
pub enum Platform {
    CommunityBridge,
    Custom,
    Github,
    IssueHunt,
    Kofi,
    Liberapay,
    OpenCollective,
    Otechie,
    Patreon,
    Tidelift,
    Other(String),
}

impl From<&str> for Platform {
    fn from(platform: &str) -> Self {
        match platform.to_ascii_uppercase().as_str() {
            "COMMUNITY_BRIDGE" => Self::CommunityBridge,
            "CUSTOM" => Self::Custom,
            "GITHUB" => Self::Github,
            "ISSUEHUNT" => Self::IssueHunt,
            "KO_FI" => Self::Kofi,
            "LIBERAPAY" => Self::Liberapay,
            "OPEN_COLLECTIVE" => Self::OpenCollective,
            "OTECHIE" => Self::Otechie,
            "PATREON" => Self::Patreon,
            "TIDELIFT" => Self::Tidelift,
            _ => Self::Other(platform.to_string()),
        }
    }
}

#[derive(Clone, Debug, Hash, Eq, PartialEq)]
pub struct Link {
    platform: Platform,
    uri: http::Uri,
}

impl Ord for Link {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        match self.platform.cmp(&other.platform) {
            std::cmp::Ordering::Equal => self.uri.to_string().cmp(&other.uri.to_string()),
            other => other,
        }
    }
}

impl PartialOrd for Link {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(&other))
    }
}

impl TryFrom<(&str, &str)> for Link {
    type Error = Error;

    fn try_from((platform, url): (&str, &str)) -> Result<Self, Self::Error> {
        let platform = platform.try_into()?;
        let mut uri: http::Uri = if url.starts_with("http") {
            url.parse()?
        } else {
            // Try https if there's no scheme
            format!("https://{}", url).parse()?
        };
        if let Platform::Github = platform {
            // fix up the URI for github sponsors 🤷
            let mut parts = uri.into_parts();
            parts.path_and_query = Some(
                format!(
                    "/sponsors{}",
                    parts
                        .path_and_query
                        .ok_or_else(|| anyhow!("Github URL missing path"))?
                        .as_str()
                )
                .as_str()
                .try_into()?,
            );
            uri = http::Uri::from_parts(parts)?;
        }
        Ok(Link { platform, uri })
    }
}

/// Try to get sources for a single package.
fn try_get_sources<'a>(package: &Package) -> Result<Vec<LinkSource>, Error> {
    let uri: http::Uri = if let Some(repo) = package.repository.as_ref() {
        repo.parse()?
    } else {
        return Ok(vec![]);
    };
    match uri.authority().map(|a| a.as_str()) {
        Some("github.com") | Some("www.github.com") => github::try_get_sources(uri),
        _ => Ok(vec![]),
    }
}

/// Get the sources for all dependencies in the workspace.
fn collect_sources(metadata: &Metadata) -> Result<HashMap<LinkSource, HashSet<PackageId>>, Error> {
    let mut source_map = HashMap::new();
    for pkg in &metadata.packages {
        if metadata.workspace_members.contains(&pkg.id) {
            // skip packages within our own workspace
            continue;
        }
        for source in try_get_sources(pkg)? {
            source_map
                .entry(source)
                .or_insert_with(HashSet::new)
                .insert(pkg.id.clone());
        }
    }
    Ok(source_map)
}

/// Turn the sources into a mapping between packages and sets of funding links.
async fn resolve_links(
    source_map: &HashMap<LinkSource, HashSet<PackageId>>,
) -> Result<HashMap<PackageId, HashSet<Link>>, Error> {
    // only one source for now, but other resolvers can add to this mapping later
    let mut resolved = HashMap::new();
    github::resolve_github_links(source_map, &mut resolved).await?;
    Ok(resolved)
}

/// Invert the mapping between packages and sets of funding links.
///
/// This allows us to group the output by unique sets of funding links.
fn invert_mapping(
    resolved: HashMap<PackageId, HashSet<Link>>,
) -> BTreeMap<BTreeSet<Link>, BTreeSet<PackageId>> {
    let mut inverted = BTreeMap::new();
    for (pkg, links) in resolved {
        let links: BTreeSet<Link> = links.into_iter().collect();
        inverted
            .entry(links)
            .or_insert_with(BTreeSet::new)
            .insert(pkg);
    }
    inverted
}

/// Print the results in a pretty tree.
///
/// TODO: support non-Unicode, perhaps add colors?
fn print_results(
    metadata: &Metadata,
    inverted: &BTreeMap<BTreeSet<Link>, BTreeSet<PackageId>>,
    num_found: usize,
) {
    println!(
        "{} (found funding links for {} out of {} dependencies)",
        metadata.workspace_root.display(),
        num_found,
        metadata.packages.len() - metadata.workspace_members.len()
    );
    let last_mapping_ix = if let Some(ix) = inverted.len().checked_sub(1) {
        ix
    } else {
        return;
    };
    for (mapping_ix, (links, pkgs)) in inverted.into_iter().enumerate() {
        let last_link_ix = links.len() - 1;
        for (link_ix, link) in links.into_iter().enumerate() {
            // first two characters of each link line
            match (mapping_ix, link_ix) {
                (0, 0) if last_mapping_ix == 0 => {
                    // first line of first and only link section
                    print!("──");
                }
                (mapping_ix, 0) if mapping_ix < last_mapping_ix => {
                    // first line of a link section
                    print!("├─");
                }
                (mapping_ix, _) if mapping_ix < last_mapping_ix => {
                    // non-first line of non-final link section
                    print!("│ ");
                }
                (mapping_ix, 0) if mapping_ix == last_mapping_ix => {
                    // first line of last link section of many
                    print!("└─");
                }
                // non-first line of final link section
                _ => print!("  "),
            }
            // second two characters of each link line
            match link_ix {
                0 if last_link_ix > 0 => {
                    // first link line of many
                    print!("┬─");
                }
                0 if last_link_ix == 0 => {
                    // first and only link line
                    print!("──");
                }
                link_ix if link_ix < last_link_ix => {
                    // non-first, non-final link line
                    print!("├─");
                }
                link_ix if link_ix == last_link_ix => {
                    // final link line of many
                    print!("└─");
                }
                _ => print!("  "),
            }
            println!(" {:?}", link.uri);
        }
        let last_pkg_ix = pkgs.len() - 1;
        for (pkg_ix, pkg) in pkgs.into_iter().enumerate() {
            if mapping_ix < last_mapping_ix {
                print!("│    ");
            } else {
                print!("     ");
            }
            if pkg_ix == last_pkg_ix {
                print!("└─");
            } else {
                print!("├─");
            }
            let pkg = &metadata[&pkg];
            println!(" {} {}", pkg.name, pkg.version);
        }
    }
}

#[tokio::main]
async fn main() -> Result<(), Error> {
    let env = envy::from_env::<args::Env>()?;
    let Opts::Fund(args) = Opts::parse();
    initialize_globals(&env, &args)?;
    let metadata = metadata::get(&args)?;
    let source_map = collect_sources(&metadata)?;
    let resolved = resolve_links(&source_map).await?;
    let num_found = resolved.len();
    let inverted = invert_mapping(resolved);
    print_results(&metadata, &inverted, num_found);
    Ok(())
}
